src/lexer.coffee

```diff
@@ -483,106 +453,79 @@ exports.Lexer = class Lexer
-    {heredoc, regex, offsetInChunk, strOffset, lexedLength} = options
-    offsetInChunk ||= 0
-    strOffset ||= 0
-    lexedLength ||= str.length
-
-    # Parse the string.
+  # This method allows us to have strings within interpolations within strings,
+  # ad infinitum.
+  matchWithInterpolations: (str, regex, end, offsetInChunk) ->
     tokens = []
-    pi = 0
-    i  = -1
-    while letter = str.charAt i += 1
-      if letter is '\\'
-        i += 1
-        continue
-      unless letter is '#' and str.charAt(i+1) is '{' and
-             (expr = @balancedString str[i + 1..], '}')
-        continue
-      # NEOSTRING is a fake token.  This will be converted to a string below.
-      tokens.push @makeToken('NEOSTRING', str[pi...i], strOffset + pi) if pi < i
-      unless errorToken
-        errorToken = @makeToken '', 'string interpolation', offsetInChunk + i + 1, 2
-      inner = expr[1...-1]
-      if inner.length
-        [line, column] = @getLineAndColumnFromChunk(strOffset + i + 2)
-        nested = new Lexer().tokenize inner, line: line, column: column, rewrite: off
-        popped = nested.pop()
-        popped = nested.shift() if nested[0]?[0] is 'TERMINATOR'
-        if len = nested.length
-          if len > 1
-            nested.unshift @makeToken '(', '(', strOffset + i + 1, 0
-            nested.push    @makeToken ')', ')', strOffset + i + 1 + inner.length, 0
-          # Push a fake 'TOKENS' token, which will get turned into real tokens below.
-          tokens.push ['TOKENS', nested]
-      i += expr.length
-      pi = i + 1
-    tokens.push @makeToken('NEOSTRING', str[pi..], strOffset + pi) if i > pi < str.length
-
-    # If regex, then return now and let the regex code deal with all these fake tokens
-    return tokens if regex
-
-    # If we didn't find any tokens, then just return an empty string.
-    return @token 'STRING', '""', offsetInChunk, lexedLength unless tokens.length
-
-    # If the first token is not a string, add a fake empty string to the beginning.
-    tokens.unshift @makeToken('NEOSTRING', '', offsetInChunk) unless tokens[0][0] is 'NEOSTRING'
+    loop
+      [strPart] = regex.exec str
 
+      # Push a fake 'NEOSTRING' token, which will get turned into a real string later.
+      tokens.push @makeToken 'NEOSTRING', strPart, offsetInChunk
+
+      str = str[strPart.length..]
+      offsetInChunk += strPart.length
+
+      break unless str[...2] is '#{'
+
+      # The `1`s are to remove the `#` in `#{`.
+      [line, column] = @getLineAndColumnFromChunk offsetInChunk + 1
+      {tokens: nested, index} =
+        new Lexer().tokenize str[1..], line: line, column: column, untilBalanced: on
+      # Skip the trailing `}`.
+      index += 1
+
+      # Remove leading and trailing `{` and `}`.
+      nested.shift()
+      nested.pop()
+
+      # Remove leading 'TERMINATOR' (if any).
+      nested.shift() if nested[0]?[0] is 'TERMINATOR'
+
+      if len = nested.length
```

Comment:
That's how I'd write it too. But here I have just moved code from one place to another. Should I change it anyway?